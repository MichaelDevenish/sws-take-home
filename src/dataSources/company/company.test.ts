
import {Company} from "./index";
import {Prisma, PrismaClient, CompanyScore, CompanyPriceClose} from "@prisma/client";
import { mockDeep, mockReset } from "jest-mock-extended";

const prismaMock = mockDeep<PrismaClient>();

const mockCompanies = [
    {
        id: 'test1',
        name: "test1",
        tickerSymbol: null,
        exchangeSymbol: null,
        uniqueSymbol: null,
        dateGenerated: null,
        securityName: null,
        exchangeCountryIso: null,
        listingCurrencyIso: null,
        canonicalUrl: null,
        uniqueSymbolSlug: null,
        scoreId: null
    },
    {
        id: 'test2',
        name: "test2",
        tickerSymbol: null,
        exchangeSymbol: null,
        uniqueSymbol: null,
        dateGenerated: null,
        securityName: null,
        exchangeCountryIso: null,
        listingCurrencyIso: null,
        canonicalUrl: null,
        uniqueSymbolSlug: null,
        scoreId: null
    }
]

describe("Test the company datasource", () => {
    beforeEach(() => {
        mockReset(prismaMock)
    })
    test("It should list companies", async () => {
        const company = new Company(prismaMock)

        prismaMock.company.findMany.mockResolvedValue(mockCompanies)
        expect(await company.getCompanies()).toMatchSnapshot()
    });

    test("It should get a company", async () => {
        const company = new Company(prismaMock)
        const companyId = 'test1'


        const companyScore = {
            id: 1,
            name: 'John Smith'
        } as unknown as Prisma.Prisma__CompanyScoreClient<CompanyScore>;

        prismaMock.company.findUnique.mockResolvedValue(mockCompanies[0])

        expect(await company.getCompany(companyId)).toMatchSnapshot()
        expect(prismaMock.company.findUnique).toHaveBeenCalledWith({where: { id: companyId }})
    });

    test("It should get a company score", async () => {
        const company = new Company(prismaMock)
        const companyId = 'test1'

        prismaMock.company.findUnique.mockReturnValue({
            companyScore: (): CompanyScore => (
                {
                    companyId: "test1",
                    dateGenerated: '',
                    id: 0,
                    dividend: 0,
                    future: 0,
                    health: 0,
                    management: 0,
                    past: 0,
                    value: 0,
                    misc: 0,
                    total: 0,
                    sentence: null
                }
            )
        } as any)

        expect(await company.getCompanyScore(companyId)).toMatchSnapshot()
        expect(prismaMock.company.findUnique).toHaveBeenCalledWith({where: { id: companyId }})
    });

    test("It should get the most recent company close price", async () => {
        const company = new Company(prismaMock)
        const companyId = 'test1'

        const companyPriceClose = jest.fn()

        companyPriceClose.mockReturnValue( [{
            date: "",
            companyId: "",
            price: 0,
            dateCreated: ""
        }])

        prismaMock.company.findUnique.mockReturnValue({
            companyPriceClose
        } as any)

        expect(await company.getCompanyPriceClose(companyId)).toMatchSnapshot()
        expect(prismaMock.company.findUnique).toHaveBeenCalledWith({where: { id: companyId }})
        expect(companyPriceClose).toHaveBeenCalledWith({
            orderBy: { date: 'desc' },
            take: 1
        })

    });

    test("It should get the all company close prices", async () => {
        const company = new Company(prismaMock)
        const companyId = 'test1'

        const companyPriceClose = jest.fn()

        companyPriceClose.mockReturnValue( [{
            date: "",
            companyId: "",
            price: 0,
            dateCreated: ""
        }, {
            date: "",
            companyId: "",
            price: 0,
            dateCreated: ""
        }])

        prismaMock.company.findUnique.mockReturnValue({
            companyPriceClose
        } as any)

        expect(await company.companyPriceCloses(companyId)).toMatchSnapshot()
        expect(prismaMock.company.findUnique).toHaveBeenCalledWith({where: { id: companyId }})
        expect(companyPriceClose).toHaveBeenCalledWith({
            orderBy: { date: 'desc' },
        })
    });
});