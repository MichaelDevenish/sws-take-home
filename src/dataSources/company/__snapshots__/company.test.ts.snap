// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the company datasource It should get a company 1`] = `
{
  "canonicalUrl": null,
  "dateGenerated": null,
  "exchangeCountryIso": null,
  "exchangeSymbol": null,
  "id": "test1",
  "listingCurrencyIso": null,
  "name": "test1",
  "scoreId": null,
  "securityName": null,
  "tickerSymbol": null,
  "uniqueSymbol": null,
  "uniqueSymbolSlug": null,
}
`;

exports[`Test the company datasource It should get a company score 1`] = `
{
  "companyId": "test1",
  "dateGenerated": "",
  "dividend": 0,
  "future": 0,
  "health": 0,
  "id": 0,
  "management": 0,
  "misc": 0,
  "past": 0,
  "sentence": null,
  "total": 0,
  "value": 0,
}
`;

exports[`Test the company datasource It should get the all company close prices 1`] = `
[
  {
    "companyId": "",
    "date": "",
    "dateCreated": "",
    "price": 0,
  },
  {
    "companyId": "",
    "date": "",
    "dateCreated": "",
    "price": 0,
  },
]
`;

exports[`Test the company datasource It should get the most recent company close price 1`] = `
{
  "companyId": "",
  "date": "",
  "dateCreated": "",
  "price": 0,
}
`;

exports[`Test the company datasource It should list companies 1`] = `
[
  {
    "canonicalUrl": null,
    "dateGenerated": null,
    "exchangeCountryIso": null,
    "exchangeSymbol": null,
    "id": "test1",
    "listingCurrencyIso": null,
    "name": "test1",
    "scoreId": null,
    "securityName": null,
    "tickerSymbol": null,
    "uniqueSymbol": null,
    "uniqueSymbolSlug": null,
  },
  {
    "canonicalUrl": null,
    "dateGenerated": null,
    "exchangeCountryIso": null,
    "exchangeSymbol": null,
    "id": "test2",
    "listingCurrencyIso": null,
    "name": "test2",
    "scoreId": null,
    "securityName": null,
    "tickerSymbol": null,
    "uniqueSymbol": null,
    "uniqueSymbolSlug": null,
  },
]
`;
